### PR TITLE
Use shared_ptr and integrated PyBehaviorModel

### PR DESCRIPTION
**Pull request by patrickhart**
The Python behavior model can now be used in BARK.

### CODE REVIEW

File: modules/world/world.hpp [hunk 1]
```diff
@@ -115,14 +117,15 @@ class World : public commons::BaseType {
...
+  WorldPtr new_world = std::make_shared<World>(*this);
   new_world->clear_all();
   for (auto agent = agents_.begin(); agent != agents_.end(); ++agent) {
-    new_world->add_agent(AgentPtr(agent->second->Clone()));
+    new_world->add_agent(
+      std::dynamic_pointer_cast<Agent>(agent->second->Clone()));
   }
   for (auto object = objects_.begin(); object != objects_.end(); ++object) {
-    new_world->add_object(ObjectPtr(object->second->Clone()));
+    new_world->add_object(object->second->Clone());
```
Comment (klemense1) [inline modules/world/world.hpp]: I would prefer if we defined object as an abstract class and derive agent and obstacle from it. Maybe we should put that into an issue?

File: modules/world/observed_world.cpp [hunk 2]
```diff
@@ -55,24 +58,28 @@ void ObservedWorld::SetupPrediction(const PredictionSettings& settings) {
...
 }
 
-std::shared_ptr<ObservedWorld> ObservedWorld::Predict(float time_span) const {
-  std::shared_ptr<ObservedWorld> next_world(ObservedWorld::Clone());
+WorldPtr ObservedWorld::Predict(float time_span) const {
+  std::shared_ptr<ObservedWorld> next_world =
+    std::dynamic_pointer_cast<ObservedWorld>(ObservedWorld::Clone());
   next_world->Step(time_span);
-  return next_world;
+  return std::dynamic_pointer_cast<World>(next_world);
```
[the remark this answers is not in the document]
Comment (patrickhart) [reply modules/world/observed_world.cpp]: @klemense1 @juloberno is it not required for the MCTS?

File: modules/models/behavior/dynamic_model/dynamic_model.cpp [hunk 3]
```diff
@@ -18,14 +18,15 @@ using dynamic::StateDefinition::Y_POSITION;
...
+using dynamic::Input;
 
 
 DynamicBehaviorModel::DynamicBehaviorModel(
   const DynamicModelPtr& dynamic_model,
   commons::Params *params) :
   BehaviorModel(params),
   dynamic_model_(dynamic_model),
-  current_action_(3),
+  current_action_(2),
```
[the remark this answers is not in the document]
Comment (patrickhart) [reply modules/models/behavior/dynamic_model/dynamic_model.cpp]: This value is resized when using the set function. Apart from this we might not need it as we can use "last_action_" from the BehaviorModel

File: modules/world/world.hpp (hunk 1 above)
Comment (patrickhart) [reply modules/world/world.hpp]: An agent is already derived from an object. For me this makes sense.
I do not think that we need to define an object as an abstract class as there are no functions that will be shared, such as "Plan"

**Pull request merged**